ccl/changefeedccl: TestChangefeedDistributionStrategy failed

ccl/changefeedccl.TestChangefeedDistributionStrategy [failed](https://tanzanite.cluster.engflow.com/invocations/default/111fa5fd-b888-46a7-8a6f-d6b0dc995945?testReportRun=11&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [a8d806658ae8118e7aca4cd5d1810b5fdc45b15c](https://github.com/cockroachdb/cockroach/commits/a8d806658ae8118e7aca4cd5d1810b5fdc45b15c):


```
=== RUN   TestChangefeedDistributionStrategy
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestChangefeedDistributionStrategy2406907249
    test_log_scope.go:81: use -show-logs to present logs inline
    changefeed_dist_test.go:445: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedDistributionStrategy2406907249
--- FAIL: TestChangefeedDistributionStrategy (677.30s)
=== RUN   TestChangefeedDistributionStrategy/locality=constrained,distribution=balanced_simple
    changefeed_dist_test.go:226: found partitions: [{2 /Table/104/1/{2-4}, /Table/104/1/2{1-2}, /Table/104/1/2{6-8}, /Table/104/1/3{1-2}, /{Table/104/1/54-Max}} {4 /Table/104/1/{8-16}, /Table/104/1/3{0-1}, /Table/104/1/{47-54}} {6 /Table/104/1/2{3-4}, /Table/104/1/{32-47}} {8 /Table/{65-104/1/2}, /Table/104/1/{4-8}, /Table/104/1/{16-21}, /Table/104/1/2{2-3}, /Table/104/1/2{4-6}, /Table/104/1/{28-30}}]
    changefeed_dist_test.go:335: range counts: [0 16 0 16 0 16 0 17]
    changefeed_dist_test.go:443: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist_test.go:443
        	Error:      	Not equal: 
        	            	expected: 65
        	            	actual  : 64
        	Test:       	TestChangefeedDistributionStrategy/locality=constrained,distribution=balanced_simple
    --- FAIL: TestChangefeedDistributionStrategy/locality=constrained,distribution=balanced_simple (143.32s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=11</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedDistributionStrategy.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35096

--- a/pkg/ccl/changefeedccl/changefeed_dist_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_dist_test.go
@@ -13,6 +13,7 @@ import (
 	"fmt"
 	"math"
 	"reflect"
+	"strings"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -21,6 +22,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
@@ -442,4 +445,73 @@ func TestChangefeedDistributionStrategy(t *testing.T) {
 		}
 		require.Equal(t, totalRanges, 64)
 	})
+}
+
+// TestDistSenderAllRangeSpans tests (*distserver).AllRangeSpans.
+func TestDistSenderAllRangeSpans(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	const nodes = 1
+	args := base.TestClusterArgs{
+		ServerArgsPerNode: map[int]base.TestServerArgs{},
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestTenantProbabilisticOnly,
+		},
+	}
+
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, nodes, args)
+	defer tc.Stopper().Stop(ctx)
+
+	node := tc.Server(0).ApplicationLayer()
+	systemLayer := tc.Server(0).SystemLayer()
+	sqlDB := sqlutils.MakeSQLRunner(node.SQLConn(t))
+	distSender := node.DistSenderI().(*kvcoord.DistSender)
+
+	tenantPrefix := ""
+	if tc.StartedDefaultTestTenant() {
+		tenantPrefix = "/Tenant/10"
+	}
+
+	// Use manual merging/splitting only.
+	tc.ToggleReplicateQueues(false)
+
+	mergeRange := func(key interface{}) {
+		err := systemLayer.DB().AdminMerge(ctx, key)
+		if err != nil {
+			if !strings.Contains(err.Error(), "cannot merge final range") {
+				t.Fatal(err)
+			}
+		}
+	}
+	getTableSpan := func(tableName string) roachpb.Span {
+		desc := desctestutils.TestingGetPublicTableDescriptor(
+			node.DB(), node.Codec(), "defaultdb", "a")
+		return desc.PrimaryIndexSpan(node.Codec())
+	}
+	getTableDesc := func(tableName string) catalog.TableDescriptor {
+		return desctestutils.TestingGetPublicTableDescriptor(
+			node.DB(), node.Codec(), "defaultdb", "a")
+	}
+
+	// Regression test for the issue in #117286.
+	t.Run("returned range does not exceed input span boundaries", func(t *testing.T) {
+		// Merge 3 tables into one range.
+		sqlDB.Exec(t, "create table a (i int primary key)")
+		sqlDB.Exec(t, "create table b (j int primary key)")
+		sqlDB.Exec(t, "create table c (k int primary key)")
+		mergeRange(getTableSpan("a").Key)
+		mergeRange(getTableSpan("b").Key)
+
+		bTableID := getTableDesc("b").GetID()
+		rangeSpans, _, err := distSender.AllRangeSpans(ctx, []roachpb.Span{getTableSpan("b")})
+		require.NoError(t, err)
+
+		// Assert that the returned range is trimmed, so it only contains spans from "b" and not the entire
+		// range containing "a" and "c".
+		require.Equal(t, 1, len(rangeSpans), fmt.Sprintf("%v", rangeSpans))
+		require.Equal(t, fmt.Sprintf("%s/Table/%d/{1-2}", tenantPrefix, bTableID), rangeSpans[0].String())
+	})
+
 }

--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -2791,8 +2791,11 @@ func (ds *DistSender) maybeIncrementErrCounters(br *kvpb.BatchResponse, err erro
 	}
 }
 
-// AllRangeSpans returns the list of all ranges that cover input spans along with the
-// nodeCountHint indicating the number of nodes that host those ranges.
+// AllRangeSpans returns a list of spans equivalent to the input list of spans
+// where each returned span represents all the keys in a range.
+//
+// For example, if the input span is {1-8} and the ranges are
+// {0-2}{2-4}{4-6}{6-8}{8-10}, then the output is {1-2}{2-4}{4-6}{6-8}{8-9}.
 func (ds *DistSender) AllRangeSpans(
 	ctx context.Context, spans []roachpb.Span,
 ) (_ []roachpb.Span, nodeCountHint int, _ error) {
@@ -2810,8 +2813,19 @@ func (ds *DistSender) AllRangeSpans(
 			if !it.Valid() {
 				return nil, 0, it.Error()
 			}
+
+			// If the range boundaries are outside the original span, trim
+			// the range.
+			startKey := it.Desc().StartKey
+			if startKey.Compare(rSpan.Key) == -1 {
+				startKey = rSpan.Key
+			}
+			endKey := it.Desc().EndKey
+			if endKey.Compare(rSpan.EndKey) == 1 {
+				endKey = rSpan.EndKey
+			}
 			ranges = append(ranges, roachpb.Span{
-				Key: it.Desc().StartKey.AsRawKey(), EndKey: it.Desc().EndKey.AsRawKey(),
+				Key: startKey.AsRawKey(), EndKey: endKey.AsRawKey(),
 			})
 			for _, r := range it.Desc().InternalReplicas {
 				replicas.Set(int(r.NodeID), 0)

